core-bundle/src/Resources/contao/elements/ContentDownload.php

```diff
@@ -140,6 +142,63 @@ protected function compile()
 		$this->Template->meta = $arrMeta;
 		$this->Template->extension = $objFile->extension;
 		$this->Template->path = $objFile->dirname;
+		$this->Template->previews = $this->getPreviews($objFile->path);
+	}
+
+	protected function getPreviews(string $path): array
+	{
+		if (!$this->showPreview)
+		{
+			return array();
+		}
+
+		$container = System::getContainer();
+		$factory = $container->get('contao.image.preview_factory');
+		$sourcePath = $container->getParameter('kernel.project_dir') . '/' . $path;
+		$builder = $container->get('contao.image.studio')->createFigureBuilder()->setSize($this->size);
+		$numberOfItems = (int) $this->numberOfItems;
+		$lightboxSize = null;
+
+		if ($this->fullsize)
+		{
+			if (!empty($GLOBALS['objPage']) && ($layoutId = $GLOBALS['objPage']->layout) && ($layout = LayoutModel::findByPk($layoutId)))
+			{
+				$lightboxSize = StringUtil::deserialize($layout->lightboxSize, true);
+			}
+
+			$builder->enableLightbox(true)->setLightboxGroupIdentifier('lb' . $this->id)->setLightboxSize($lightboxSize);
+		}
+
+		try
+		{
+			$lightboxPreviews = array();
+			$previews = $factory->createPreviews($sourcePath, $factory->getPreviewSizeFromImageSize($this->size), $numberOfItems ?: PHP_INT_MAX);
+			$previews = \is_array($previews) ? array_values($previews) : iterator_to_array($previews, false);
+
+			if ($this->fullsize)
+			{
+				$lightboxPreviews = $factory->createPreviews($sourcePath, $factory->getPreviewSizeFromImageSize($lightboxSize), $numberOfItems ?: PHP_INT_MAX);
+				$lightboxPreviews = \is_array($lightboxPreviews) ? array_values($lightboxPreviews) : iterator_to_array($lightboxPreviews, false);
+			}
+
+			foreach ($previews as $index => $preview)
+			{
+				$builder->fromImage($preview);
+
+				if (!empty($lightboxPreviews[$index]))
+				{
+					$builder->setLightboxResourceOrUrl($lightboxPreviews[$index]);
+				}
+
+				$previews[$index] = $builder->build()->getLegacyTemplateData();
```

Comment:
```suggestion
				$previews[$index] = $builder->build();
```

No need to use legacy data here I think.

Reply:
Done in 08140cb31f376e1d26dc395405ac2ef81f564409